tensorflow_recommenders_addons/dynamic_embedding/core/kernels/hkv_hashtable_op_gpu.cu.cc

```diff
@@ -0,0 +1,950 @@
+        CUDA_CHECK(cudaMemcpy((void*)d_keys, (void*)keys.tensor_data().data(),
+                              sizeof(K) * len, cudaMemcpyDefault));
+      } else {
+        d_keys = (K*)keys.tensor_data().data();
+      }
+      cudaPointerAttributes values_attr;
+      CUDA_CHECK(cudaPointerGetAttributes(&values_attr,
+                                          (void*)values.tensor_data().data()));
+      if (values_attr.type != cudaMemoryTypeDevice) {
+        CUDA_CHECK(cudaMallocManaged((void**)&d_values,
+                                     sizeof(V) * runtime_dim_ * len));
+        CUDA_CHECK(
+            cudaMemcpy((void*)d_values, (void*)values.tensor_data().data(),
+                       sizeof(V) * runtime_dim_ * len, cudaMemcpyDefault));
+      } else {
+        d_values = (V*)values.tensor_data().data();
+      }
+      {
+        mutex_lock l(mu_);
+        try {
+          table_->clear(stream);
+          table_->upsert((const K*)d_keys, (const V*)d_values, len, stream);
+          CUDA_CHECK(cudaStreamSynchronize(stream));
+        } catch (std::runtime_error& e) {
+          return Status(tensorflow::error::INTERNAL, e.what());
+        }
+      }
+      if (keys_attr.type != cudaMemoryTypeDevice) {
+        CUDA_CHECK(cudaFree(d_keys));
+      }
+      if (values_attr.type != cudaMemoryTypeDevice) {
+        CUDA_CHECK(cudaFree(d_values));
+      }
+    }
+    return Status::OK();
+  }
+
+  Status ExportValues(OpKernelContext* ctx) override {
+    size_t len = 0;
+    int64 size = 0;
+
+    const size_t offset = 0;
+
+    Tensor* keys;
+    Tensor* values;
+
+    size_t* d_dump_counter = nullptr;
+    auto stream = ctx->eigen_device<GPUDevice>().stream();
+
+    {
+      tf_shared_lock l(mu_);
+      len = table_->get_capacity();
+      size = (int64)table_->get_size(stream);
+      CUDA_CHECK(cudaStreamSynchronize(stream));
+    }
+
+    CUDA_CHECK(cudaMallocAsync(&d_dump_counter, sizeof(size_t), stream));
+    CUDA_CHECK(cudaMemsetAsync(d_dump_counter, 0, sizeof(size_t), stream));
+
+    AllocatorAttributes attr;
+    // attr.set_gpu_compatible(true);
+    // attr.set_nic_compatible(true);
+    attr.set_on_host(false);
+
+    TF_RETURN_IF_ERROR(
+        ctx->allocate_output("keys", TensorShape({(size)}), &keys, attr));
+    TF_RETURN_IF_ERROR(ctx->allocate_output(
+        "values", TensorShape({size, (int64)runtime_dim_}), &values, attr));
+    if (size) {
+      tf_shared_lock l(mu_);
+      try {
+        table_->dump((K*)keys->flat<K>().data(),
+                     (V*)(values->matrix<V>().data()), offset, len,
+                     d_dump_counter, stream);
+        CUDA_CHECK(cudaStreamSynchronize(stream));
+      } catch (std::runtime_error& e) {
+        return Status(tensorflow::error::INTERNAL, e.what());
+      }
+    }
+    CUDA_CHECK(cudaFreeAsync(d_dump_counter, stream));
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+    return Status::OK();
+  }
+
+  Status ExportValuesWithMetas(OpKernelContext* ctx) {
```

Comment:
Hi @LinGeLin , It would be better to change all of the Meta to Score. In the HKV, we do have change it permarnently.

Reply:
Can I update it later? Because we are currently using HKV bate6, and the changes to scores are in bate7. The code for setting the thrust custom allocator has not been released yet, so I would like to update it later. Also, we haven't registered the ExportMate operator yet, so we can consider it as a feature update in the future. Is that okay?

Reply:
If the user API is not named xxMetaxx, it's OK.